src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -106,7 +106,7 @@ internal sealed class TupleTypeSymbol : WrappedNamedTypeSymbol
             var constructedType = Create(underlyingType, elementNames, errorPositions, locationOpt, elementLocations);
             if (shouldCheckConstraints && diagnostics != null)
             {
-                constructedType.CheckConstraints(compilation.Conversions, syntax, elementLocations, compilation, diagnostics);
+                constructedType.CheckConstraints(compilation.Conversions, compilation.Conversions.IncludeNullability, syntax, elementLocations, compilation, diagnostics);
```

Comment:
>compilation.Conversions.IncludeNullability [](start = 74, length = 42)

Should we simply pass ```false``` here?

Reply:
Don't believe so. The code goes through the process of validating constraints here hence it woudl seem that nullable should be checked as well (as it was before my change)., Is the expectation that the nullable constraints are always checked elsewhere, say NullableWalker, for method bodies?

---
In reply to: [255382791](https://github.com/dotnet/roslyn/pull/33261#discussion_r255382791) [](ancestors = 255382791)

Reply:
@AlekseyTs: @cston and I have chatted about this and I'm a bit stuck on how to proceed.

On one hand the compiler tends to special case tuple and it's questionable if tuple constraints should matter at all. But the code clearly has the intent of validating tuple constaints. Hence it seems that we should consider tuple constraints in this scenario and I can't see an obvious reason for excluding nullability ones. 

As an experiment I passed `false` as you suggested and no tests failed. But that possibly means we're just not validating the correct scenario here. A test where the tuple had say a `class` constraint for every type parameter could just not be present to catch this.

Can you elaborate a bit on why you think `false` is the right check here?

Reply:
> Don't believe so. The code goes through the process of validating constraints here hence it woudl seem that nullable should be checked as well (as it was before my change)., Is the expectation that the nullable constraints are always checked elsewhere, say NullableWalker, for method bodies?

I doubt nullable aspect of constraints was checked here before. Is ```compilation.Conversions``` ever checks nullable? I think that your intuition is probably correct that for some cases nullbility aspect should be checked here, but probably not for all cases. For example, when this method is called from ```BindTupleExpression```, it feels like the check for the nullability aspect should be postponed until nullable analysis, when we know nullability of expressions involved. Deconstruction is the other example. It feels like the change in this function should be scenario and unit-test driven. It would be fine to follow up on this separately.  

---
In reply to: [255648877](https://github.com/dotnet/roslyn/pull/33261#discussion_r255648877) [](ancestors = 255648877,255382791)

Reply:
> Can you elaborate a bit on why you think ```false``` is the right check here? 

As I mentioned in my other response to this thread, I believe that would be consistent with the previous behavior. This method is never called from nullable analysis. I think that it would be good to back this change with targeted unit-tests that cover every call site for this function.

---
In reply to: [255666778](https://github.com/dotnet/roslyn/pull/33261#discussion_r255666778) [](ancestors = 255666778)

Reply:
> I doubt nullable aspect of constraints was checked here before. Is compilation.Conversions ever checks nullable? 

This is the root of my confusion. I had mistakenly believed that `compliation.Conversions.IncludeNullability` was toggled based on C# 8.0 Instead looking now that is a case where it is essentially always `false`. Agree `false` is correct here. 

> . It feels like the change in this function should be scenario and unit-test driven.

Agree. Want to make sure I get this tested. Prefer to just take care of it now vs. waiting till later if possible.


---
In reply to: [255676114](https://github.com/dotnet/roslyn/pull/33261#discussion_r255676114) [](ancestors = 255676114,255648877,255382791)

Reply:
Dug into this a bit and decided to pull out this validation to a separate PR. The validation of tuple constraints is now done properly in method signatures, which is the intent of this PR. The problem is the validation within method bodies. Will follow up on that separately

[https://github.com/dotnet/roslyn/issues/33303](https://github.com/dotnet/roslyn/issues/33303)

---
In reply to: [255689761](https://github.com/dotnet/roslyn/pull/33261#discussion_r255689761) [](ancestors = 255689761,255676114,255648877,255382791)